Sources/GRPCCore/Metadata.swift

```diff
@@ -14,5 +14,267 @@
+import Foundation
+
+/// A collection of metadata key-value pairs, found in RPC streams.
+/// A key can have multiple values associated to it.
+/// Values can be either strings or binary data, in the form of `Data`.
+/// - Note: Binary values must have keys ending in `-bin`, and this will be checked when adding pairs.
+public struct Metadata: Sendable, Hashable {
+
+  /// A metadata value. It can either be a simple string, or binary data.
+  public enum MetadataValue: Sendable, Hashable {
+    case string(String)
+    case binary(Data)
+  }
+
+  /// A metadata key-value pair.
+  public struct MetadataKeyValue: Sendable, Hashable {
+    internal let key: String
+    internal let value: MetadataValue
+
+    /// Constructor for a metadata key-value pair.
+    /// - Parameters:
+    ///   - key: The key for the key-value pair.
+    ///   - value: The value to be associated to the given key. If it's a binary value, then the associated
+    ///   key must end in `-bin`, otherwise, this method will produce an assertion failure.
+    init(key: String, value: MetadataValue) {
+      if case .binary = value {
+        assert(key.hasSuffix("-bin"), "Keys for binary values must end in -bin")
+      }
+      self.key = key
+      self.value = value
+    }
+  }
+
+  private let lockedElements: LockedValueBox<[MetadataKeyValue]>
+  private var elements: [MetadataKeyValue] {
+    get {
+      self.lockedElements.withLockedValue { $0 }
+    }
+    set {
+      self.lockedElements.withLockedValue { $0 = newValue }
+    }
+  }
+
+  /// The Metadata collection's capacity.
+  public var capacity: Int {
+    self.elements.capacity
+  }
+
+  /// Initialize an empty Metadata collection.
+  public init() {
+    self.lockedElements = .init([])
+  }
+
+  public static func == (lhs: Metadata, rhs: Metadata) -> Bool {
+    lhs.elements == rhs.elements
+  }
+
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(self.elements)
+  }
+
+  /// Reserve the specified minimum capacity in the collection.
+  /// - Parameter minimumCapacity: The minimum capacity to reserve in the collection.
+  public mutating func reserveCapacity(_ minimumCapacity: Int) {
+    self.elements.reserveCapacity(minimumCapacity)
+  }
+
+  /// Add a new key-value pair, where the value is a string.
+  /// - Parameters:
+  ///   - key: The key to be associated with the given value.
+  ///   - stringValue: The string value to be associated with the given key.
+  public mutating func add(key: String, stringValue: String) {
+    self.add(key: key, value: .string(stringValue))
+  }
+
+  /// Add a new key-value pair, where the value is binary data, in the form of `Data`.
+  /// - Parameters:
+  ///   - key: The key to be associated with the given value. Must end in `-bin`.
+  ///   - binaryValue: The `Data` to be associated with the given key.
+  public mutating func add(key: String, binaryValue: Data) {
+    self.add(key: key, value: .binary(binaryValue))
+  }
```

Comment:
I don't love the `stringValue`/`binaryValue` naming here. Dictionary inverts the order and has `updateValue(_:forKey:)` so I think `add{String,Binary,Value}(_:forKey:)` would be the better spelling here. (same applies to the replace-or-add API)

 I'm also wondering if we actually need these overloads at all. The common case will be adding strings, so I'm wondering if adding `ExpressibleByStringLiteral` and `ExpressibleByStringInterpolation` conformance to `MetadataValue` gets us most of the benefit without having to carry the extra API. WDYT? (We should add the conformances either way.)

Reply:
> I'm also wondering if we actually need these overloads at all. 

Having them would slightly simplify adding binary values (as we'd avoid having to wrap the `[UInt8]`s in `.binary(...)`, but I agree it's quite minor and probably not a very common use case. I don't really have a strong opinion here, so if we prefer to have a smaller API surface I can get rid of the overloads.

Reply:
Let's keep them. There's not much overhead and there's definitely a usability win.